orderer/common/cluster/clusterservice.go

```diff
@@ -0,0 +1,262 @@
+	request, err := stream.Recv()
+	if err == io.EOF {
+		return err
+	}
+	if err != nil {
+		s.Logger.Warningf("Stream read from %s failed: %v", addr, err)
+		return err
+	}
+	if request == nil {
+		return errors.Errorf("Request message is nil")
+	}
+	if s.StepLogger.IsEnabledFor(zap.DebugLevel) {
+		nodeName := commonNameFromContext(stream.Context())
+		s.StepLogger.Debugf("Received message from %s(%s): %v", nodeName, addr, clusterRequestAsString(request))
+	}
+
+	exp.checkExpiration(time.Now(), channel)
+
+	if submitReq := request.GetNodeTranrequest(); submitReq != nil {
+		return s.handleSubmit(submitReq, stream, addr, channel)
+	} else if clusterConReq := request.GetNodeConrequest(); clusterConReq != nil {
+		conReq := &orderer.ConsensusRequest{
+			Channel:  channel,
+			Payload:  clusterConReq.Payload,
+			Metadata: clusterConReq.Metadata,
+		}
+		return s.Dispatcher.DispatchConsensus(stream.Context(), conReq)
+	}
+	return errors.Errorf("Message is neither a Submit nor Consensus request")
+}
+
+func (s *ClusterService) handleSubmit(request *orderer.NodeTransactionOrderRequest, stream ClusterStepStream, addr string, channel string) error {
+	submitReq := &orderer.SubmitRequest{
+		Channel:           channel,
+		LastValidationSeq: request.LastValidationSeq,
+		Payload:           request.Payload,
+	}
+	err := s.Dispatcher.DispatchSubmit(stream.Context(), submitReq)
+	if err != nil {
+		s.Logger.Warningf("Handling of Submit() from %s failed: %v", addr, err)
+		return err
+	}
+	return err
+}
+
+func (s *ClusterService) initializeExpirationCheck(stream orderer.ClusterNodeService_StepServer, endpoint, nodeName string) *certificateExpirationCheck {
+	expiresAt := time.Time{}
+	cert := util.ExtractCertificateFromContext(stream.Context())
+	if cert != nil {
+		expiresAt = cert.NotAfter
+	}
+
+	return &certificateExpirationCheck{
+		minimumExpirationWarningInterval: s.MinimumExpirationWarningInterval,
+		expirationWarningThreshold:       s.CertExpWarningThreshold,
+		expiresAt:                        expiresAt,
+		endpoint:                         endpoint,
+		nodeName:                         nodeName,
+		alert: func(template string, args ...interface{}) {
+			s.Logger.Warningf(template, args...)
+		},
+	}
+}
+
+func (c *ClusterService) ConfigureNodeCerts(channel string, newNodes []NodeIdentity) error {
+	if c.MembershipByChannel == nil {
+		return errors.New("Nodes cert storage is not initialized")
+	}
+
+	c.Logger.Infof("Updating nodes identity, channel: %s, nodes: %v", channel, newNodes)
+
+	c.Lock.Lock()
+	defer c.Lock.Unlock()
+
+	channelCert := c.MembershipByChannel[channel]
+	if channelCert != nil {
+		c.Logger.Debugf("Refreshing the %s channel nodes identities", channel)
+		delete(c.MembershipByChannel, channel)
+	}
+	channelCert = make(map[uint64]*msp.SerializedIdentity)
+	for _, nodeIdentity := range newNodes {
+		sID := &msp.SerializedIdentity{}
+		proto.Unmarshal(nodeIdentity.Identity, sID)
```

Comment:
We should have a mapping of MSPID and Identity, not SerializedIdentity, because that's how we encode in the channel config.